### ISSUE TITLE
[SPARK-39792][SQL] Add DecimalDivideWithOverflowCheck for decimal average

### ISSUE THREAD

**Issue by amahussein** [NVIDIA/spark-rapids#6076]
**Describe the bug**
Add a new expression `DecimalDivideWithOverflowCheck` to replace the previous CheckOverflowInSum + Divide + Cast.

**Expected behavior**
If the result data type is decimal, the Average will first calculate the result using the default precision and scale of divide, then cast to the result data type. We should do calculate and return the result data type directly so that we can avoid the precision loss. It can also save one unnecessary cast.
And for the overflow check, we should check the result of divide whether overflow instead of the dividend.

**Additional context**
https://github.com/apache/spark/commit/46071a9caa

**Comment by razajafri** [comment]
@amahussein is this is performance improvement or a build/test failure?

**Comment by NVnavkumar** [comment]
This looks like a Spark 3.4 incompatibility, but there might be no test coverage specifically for this edge case (potential loss of precision)

**Comment by andygrove** [comment]
I implemented a version of the test `SPARK-24957: average with decimal followed by aggregation returning wrong result` from the Spark PR linked to in this issue's description.

When I ran this test against Spark 3.3.2, both CPU  and GPU produced `11.9999999994857142860000`.

When I ran this test against Spark 3.4.0, both CPU  and GPU produced `11.9999999994857142857143`.

These values match the before and after values in the Spark PR. In other words, I could not reproduce the loss of precision using this test. I will continue to investigate.
